### ISSUE TITLE
Changes to Index.jsp to add rel="noopener noreferrer" next to target="_blank" in the Youtube link

### ISSUE THREAD

**Issue by nbiswal** [TEAMMATES/teammates#8183]
Hello Teammates,

We are a group of 4 people working on this Open Source Software project(Teammates) as part of our subject "Software Assurance". We are mainly focusing on the security aspects of this project.

While exploring for the security guidelines followed in the project we could see in the index.jsp (location - src/main/webapp/index.jsp) for the YouTube link "https://www.youtube.com/embed/mDtfmNmRwBM?autoplay=1&rel=0", **target="_blank"** is mentioned but **rel="noopener noreferrer"** is not mentioned.And as we know if  rel="noopener noreferrer" is not mentioned with target="_blank", then there is a possibility of reverse tab-nabbing.

Below is an attached screenshot of the suggested change that needs to be done in index.jsp().

![changed index](https://user-images.githubusercontent.com/26098493/33299367-dccb0d1e-d3b0-11e7-91ce-68a118b821e5.JPG)

Kindly get back in case of any concerns. 

Thanks
Niluja Biswal,Neeharica Kodali,Durgabhavani Peta,Kasi Viswanath
University Of Nebraska Omaha,Nebraska

**Comment by LiHaoTan** [comment]
Thanks for informing!

I'm no security expert so let's wait for others to chip in. But as far as I understand, this is not something urgent because Youtube is unlikely to be doing anything funny.

I'm not sure if there is any consensus anywhere to implement this though because it does not fully address the whole class of attacks.

> Unfortunately, we believe that this class of attacks is inherent to the current design of web browsers and can't be meaningfully mitigated by any single website; in particular, clobbering the window.opener property limits one of the vectors, but still makes it easy to exploit the remaining ones. 

https://sites.google.com/site/bughunteruniversity/nonvuln/phishing-with-window-opener

An example of a similar attack is highlighted as follows:
> I mentioned this on another related thread: https://news.ycombinator.com/item?id=11554080. In short, the solution here only closes one of potentially many other similar attacks: http://lcamtuf.coredump.cx/switch/
>
> The attack I linked to originates from the malicious site and links to the trusted site (the reverse of the attack in this post). There is nothing a site can do to prevent this since there is no way for a linked to site to prevent the linking site from getting a window reference to it. So, imagine a scenario like this:
> * trusted site implements the guidance here and links to malicious site.
> * The malicious site, detecting that they can't get a reference to "window.opener" immediately opens a new tab back to the trusted site (maybe to the login page if the site has any logout CSRF issues).
>* The user is slightly confused, but they were just on the trusted site, so it doesn't feel too strange.
>* If the user is super savvy, the look up at the URL bar and are assured that they actually are back in the trusted site (possibly staring at a login prompt).
>* the attacker has a reference to the window they opened back to the trusted site. They set a timer for a couple seconds (like the attack I referenced above). After a few seconds they change the trusted site to load a malicious site and/or malicious data URL.
>* user "logs in to the trusted site" and gives up their creds.

https://news.ycombinator.com/item?id=11631292

**Comment by chowyb** [comment]
@nbiswal yes, this seems to be a valid issue. The idea to use `rel="noopener noreferrer"` was first introduced in #6367, but perhaps the link to the video was unintentionally left out.

If you wish, you may want to submit a PR rectifying this issue, and we'd be happy to review the PR :)

**Comment by wkurniawan07** [comment]
@nbiswal there's in fact quite a number of places where such inclusion is missing (i.e. this is just one of the many). You're welcome to patch as many as you can find.

@LiHaoTan very insightful link. One thing though, I notice you have this tendency for "all or nothing" approach. Patching one source of error of course doesn't help in patching everything else, but it's better than patching none, right?

**Comment by nbiswal** [comment]
Thanks @wkurniawan07 @chowyb. I would be providing a pull request soon with the required changes and I would try to add the rel="noopener noreferrer" wherever its missing.

**Comment by LiHaoTan** [comment]
@wkurniawan07 right, I realized I had the tendency to lean towards an "all or nothing" approach sometimes and it's not very good. I am definitely trying to address it 😅 .

This doesn't quite apply here though. What I meant here is that if we do pick the approach taken here, we need to make sure we are aware that we are not actually fixing the problem completely. However, our code and documentation give the impression that we are actually resolving the security issue.

That is to say, if we fix the current issue, we need to re-frame possibly the documentation to say that the issue is partially addressed.

I am also partly influenced by how this security issue seems to be largely ignored by many, e.g. WordPress only recently started adding this.

Now I am thinking this is a little harder than it seems because for a high priority security issue, we need to quickly address it and I suspect it is not the case here. Then, we still have to carefully evaluate if we actually address the security issue, but how far we are going to go depends on the severity, complexity, and priority.

Ultimately, the question that remains is why for example WordPress only recently decided it is important and why browsers are not going to do anything about it. Yes, they probably have another issue to consider, that is backward compatibility.

In conclusion, I am definitely leaning towards adding `rel="noopener noreferrer"`, but it must be attached with the limitation in some form like documentation. The problem that remains is if I missed out considering any other issue on top of how this issue is ignored largely for a long time.

**Comment by wkurniawan07** [comment]
> we need to make sure we are aware that we are not actually fixing the problem completely

> if we fix the current issue, we possibly need to re-frame possibly the documentation to say that the issue is partially addressed

> it must be attached with the limitation in some form like documentation

@LiHaoTan You can try to fix it, although I personally am quite wary of inclusion of such. If anything, it's because security is a very grey area and there's always new vulnerabilities coming up on a regular basis. It's a lose-lose situation right now.
- Not including: we may give an impression that there's only one such vulnerability.
- Including: what do you want to achieve by such inclusion? Sure, readers will know that such vulnerabilities exist, but what do you expect afterwards? If it does not translate into a concrete action ~~(e.g. do not use `target="_blank"` at all)~~, it does not value add much, right? If it's just to give information, a "coding standards" document is not exactly the right place for it. It fits into a "coding standards" document only if you _include and follow_.<br>And to open another can of worms: even if you include and follow, how certain are you that there are no other vulnerabilities lurking out there? We're getting back to the problem of "not including".

**Comment by LiHaoTan** [comment]
@wkurniawan07 I think I agree with your points and it's also part of why it is difficult. New vulnerabilities always pop up, e.g. like the recent [IDN homograph attack](https://en.wikipedia.org/wiki/IDN_homograph_attack)

Basically, I am not sure if we should ever claim that we are fixing a security issue unless it is verified by industry experts, e.g. best practices for handling CSRF attacks. The problem with this issue is that I can't find any recommendations anywhere, it may exist but I can't find it.

Then we are left on our own to address the issue, and the coding standards kind of gives the impression we are fixing the vulnerability. I don't think we should claim that because the only thing we do know is how it prevents `window.opener` from being set and mitigate some possible security issues.

Surely saying that it partially addresses the issue still gives an actionable way to handle the issue?

It's true that it probably shouldn't exist in a coding standards document still but we don't really have a security guidelines document,  so I was thinking until we have one at least that our coding standards need to reflect the security fix accurately?

**Comment by nbiswal** [comment]
Hello everyone,
While exploring the issue I could find the corresponding CWE-ID for the same.Attached are the details for your reference.

CWE-1022: Improper Restriction of Cross-Origin Permission to window.opener.location (https://cwe.mitre.org/data/definitions/1022.html)

Thanks!

**Comment by LiHaoTan** [comment]
@nbiswal That looks like a great source though I am unable to verify how good the source is.

I went across a few different browsers to replicate the security issue and am unable to replicate the security issue. It's as though this issue has already been fixed by modern browsers and I'm not sure if it's applicable anymore.

# Results
## Tested browsers without cross-origin `window.opener` vulnerability
Chrome 62.0.3202.94
Firefox 57.0, 46.0.1, 34.0, 33.0
Safari 11.0.1

## Tested Browsers with cross-origin` window.opener` vulnerability
Chrome 48.0.2564.109

---
Looks we need some confirmation on this issue.

**Comment by LiHaoTan** [comment]
I have found the patch for the behavior:
https://chromium.googlesource.com/chromium/src/+/b2607b8c1ec79075a2586c87aed3fce9645b561d

Some additional details:
https://html.spec.whatwg.org/multipage/history.html#dom-location-assign

It seems that the behavior has been flip-flopping on Chromium and if I remember correctly it is because the specs on WhatWG has been changing? I will be updating this comment to provide more context soon.

I suspect we may be able to revert all `rel="noopener no referrer"` very soon because this patch probably has existed for a sufficiently long time, around 6 months since the release of Chrome `57.0.2987`, and the patch is done in between `57.0.2963.0` and `57.0.2964.0`.

Diff between Chromium 57.0.2963.0 and Chromium 57.0.2964.0
https://chromium.googlesource.com/chromium/src/+log/a46975912f4918139134ab42b1ef9259b078fa98..3b55afac722a917f6a69200262a8c466b78476b4/?s=1fe146a0bb60013b3f74c4db7f8bb2409e80eb90

**Comment by wkurniawan07** [comment]
@LiHaoTan one problem with your argument is you're assuming all users are using latest Chrome...

**Comment by LiHaoTan** [comment]
@wkurniawan07 Across the major browsers, the vulnerability pretty much only exists on Chrome.

> location.assign shouldn't be cross-origin accessible, but
we need assignOriginSafeMethodGetter to simulate the Incumbent
realm.  See the comment in Location.idl for details.

> Other browsers' behavior is:
FF, Safari = location.assign is not cross-origin accessible.
Edge = cross-origin representation of |location| is not
supported.
So, no one seems supporting location.assign on cross-origin
except for Blink.

**Comment by wkurniawan07** [comment]
Resolved in V7.
